File: 14.longest-common-prefix/csharp/longest_common_prefix.cs
﻿using System;

namespace LeetCode
{
    class Solution
    {
        static void Main(string[] args)
        {
            Test test_01 = new Test{ Strs = new string[] { "flower","flow","flight" } };
            Test test_02 = new Test{ Strs = new string[] { "dog","racecar","car" } };

            Solution solution = new Solution();

            string result_01 = solution.LongestCommonPrefix(test_01.Strs);
            string result_02 = solution.LongestCommonPrefix(test_02.Strs);

            Console.WriteLine();
            Console.WriteLine("Expected: 'fl'" + "\n" + "Output: " + result_01);
            Console.WriteLine();
            
            Console.WriteLine("Expected: ''" + "\n" + "Output: " + result_02);
            Console.WriteLine();
        }

        public string LongestCommonPrefix(string[] strs)
        {
            if (! IsValid(strs)) {
                return "";
            }

            int low = 1;
            int high = GetMinLen(strs);
            int mid;

            while (low <= high) {
                mid = (low + high) / 2;
                
                if (IsCommonPrefix(mid, strs)) {
                    low = mid + 1;
                    continue;
                }

                high = mid - 1;
            }

            int result = (low + high) / 2;
            return strs[0].Substring(0, result);
        }

        public bool IsValid(string[] strs)
        {
            return ! (strs == null || strs.Length == 0);
        }

        public int GetMinLen(string[] strs)
        {
            int minLen = strs[0].Length;

            for (int i = 0; i < strs.Length; i++) {
                if (strs[i].Length < minLen) {
                    minLen = strs[i].Length;
                }
            }

            return minLen;
        }

        public bool IsCommonPrefix(int mid, string[] strs)
        {
            string prefix = strs[0].Substring(0, mid);

            for (int i = 1; i < strs.Length; i++) {
                if (! strs[i].StartsWith(prefix)) {
                    return false;
                }
            }

            return true;
        }
    }

    class Test
    {
       public string[] Strs { get; set; }
    }
}

File: 14.longest-common-prefix/java/LongestCommonPrefix.java
class Main
{
    public static void main(String[] args)
    {
        Test test_01 = new Test(new String[] { "flower", "flow", "flight" });
        Test test_02 = new Test(new String[] { "dog", "rececar", "car" });

        Solution solution = new Solution();

        String result_01 = solution.longestCommonPrefix(test_01.strs);
        String result_02 = solution.longestCommonPrefix(test_02.strs);

        System.out.println();
        System.out.println("Expected: 'fl'" + "\n" + "Output: " + result_01);
        System.out.println();
        
        System.out.println("Expected: ''" + "\n" + "Output: " + result_02);
        System.out.println();
    }
}

class Solution
{
    public String longestCommonPrefix(String[] strs)
    {
        if (! isValid(strs)) {
            return "";
        }

        int low = 1;
        int high = getMinLen(strs);
        int mid;

        while (low <= high) {
            mid = (low + high) / 2;

            if (isCommonPrefix(mid, strs)) {
                low = mid + 1;
                continue;
            }

            high = mid - 1;
        }

        int result = (low + high) / 2;
        return strs[0].substring(0, result);
    }

    public Boolean isValid(String[] strs)
    {
        return (! (strs == null || strs.length == 0)) ? true : false;
    }

    public int getMinLen(String[] strs)
    {
        int minLen = strs[0].length();

        for (int i = 1; i < strs.length; i++) {
            if (strs[i].length() < minLen) {
                minLen = strs[i].length();
            }
        }

        return minLen;
    }

    public Boolean isCommonPrefix(int mid, String[] strs)
    {
        String prefix = strs[0].substring(0, mid);
        
        for (int i = 1; i < strs.length; i++) {
            if (! strs[i].startsWith(prefix)) {
                return false;
            }
        }

        return true;
    }
}

class Test
{
    String[] strs;

    public Test(String[] strs) {
        this.strs = strs;
    }
}
